cudax/include/cuda/experimental/__device/all_devices.cuh

```diff
@@ -0,0 +1,244 @@
+  using reference         = device;
+  using iterator_category = ::cuda::std::random_access_iterator_tag;
+  using iterator_concept  = ::cuda::std::random_access_iterator_tag;
+
+  //! @brief Proxy object used as the return value of `operator->`
+  struct __proxy
+  {
+    _CCCL_NODISCARD const device* operator->() const noexcept
+    {
+      return &__dev_;
+    }
+
+    device __dev_;
+  };
+
+  _CCCL_NODISCARD constexpr device operator*() const noexcept
+  {
+    return device(__id_);
+  }
+
+  _CCCL_NODISCARD constexpr device operator[](int __offset) const noexcept
+  {
+    return device(__id_ + __offset);
+  }
+
+  _CCCL_NODISCARD constexpr __proxy operator->() const noexcept
+  {
+    return __proxy{device(__id_)};
+  }
+
+  constexpr iterator& operator++() noexcept
+  {
+    ++__id_;
+    return *this;
+  }
+
+  constexpr iterator operator++(int) noexcept
+  {
+    auto tmp = *this;
+    ++__id_;
+    return tmp;
+  }
+
+  constexpr iterator& operator--() noexcept
+  {
+    --__id_;
+    return *this;
+  }
+
+  constexpr iterator operator--(int) noexcept
+  {
+    auto tmp = *this;
+    --__id_;
+    return tmp;
+  }
+
+  constexpr iterator& operator+=(int __offset) noexcept
+  {
+    __id_ += __offset;
+    return *this;
+  }
+
+  _CCCL_NODISCARD constexpr iterator operator+(int __offset) noexcept
+  {
+    return iterator(__id_ + __offset);
+  }
+
+  constexpr iterator& operator-=(int __offset) noexcept
+  {
+    __id_ -= __offset;
+    return *this;
+  }
+
+  _CCCL_NODISCARD constexpr iterator operator-(int __offset) noexcept
+  {
+    return iterator(__id_ - __offset);
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr iterator operator+(int __lhs, const iterator& __rhs) noexcept
+  {
+    return iterator(__lhs + __rhs.__id_);
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr int operator-(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ - __rhs.__id_;
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr bool operator==(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ == __rhs.__id_;
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr bool operator!=(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ != __rhs.__id_;
+  }
+
+private:
+  friend struct all_devices;
+
+  constexpr explicit iterator(int __id) noexcept
+      : __id_(__id)
+  {}
+
+  int __id_ = 0;
+};
+
+_CCCL_NODISCARD inline constexpr device all_devices::operator[](size_type __id_) const noexcept
+{
+  assert(__id_ < size());
+  return device(static_cast<int>(__id_));
+}
+
+_CCCL_NODISCARD inline all_devices::size_type all_devices::size() const
+{
+  size_type __size = __size_.load(_CUDA_VSTD::memory_order_relaxed);
+  if (__size == ~0UL)
+  {
+    int __count = 0;
+    _CCCL_TRY_CUDA_API(::cudaGetDeviceCount, "failed to get the count of CUDA devices", &__count);
+    __size = static_cast<size_type>(__count);
+    __size_.store(__size, _CUDA_VSTD::memory_order_relaxed);
+  }
+  return __size;
```

Comment:
Important: this code seems fishy. Even though we somewhat have a critical section, the first load is not an `acquire` and the last store is not a `release`. Maybe this is fine, because two racing threads could execute the critical section concurrently. I am not sure how the sequentically consistent (SC) store (`__size = static_cast<size_type>(__count);`) adds to this mix. AFAIU, the SC store also has release semantics, but there is no `acquire` or SC load to synchronize with. Also, the relaxed store after the SC store seems redundant then.

I wonder whether we should just let the compiler handle the first touch initialization using:
```C++
  static size_type size = []{
    int __count = 0;
    _CCCL_TRY_CUDA_API(::cudaGetDeviceCount, "failed to get the count of CUDA devices", &__count);
    return count;
  }();
  return size;
```
And drop the `__size_` member.